File: content/LLD/Reference-Code/23. Chain of Responsibility Pattern/main.py
from approvers import CFO, Director, Manager, TeamLead
from handler import ExpenseRequest

team_lead = TeamLead()
manager = Manager()
director = Director()
cfo = CFO()

team_lead.set_next(manager).set_next(director).set_next(cfo)

requests = [
    ExpenseRequest(500, "Office supplies"),
    ExpenseRequest(5_000, "Team offsite"),
    ExpenseRequest(30_000, "New servers"),
    ExpenseRequest(100_000, "Acquisition deal"),
]

for req in requests:
    print(team_lead.handle(req))

File: content/LLD/Reference-Code/23. Chain of Responsibility Pattern/Main.java
import java.util.List;

public class Main {
    public static void main(String[] args) {
        TeamLead teamLead = new TeamLead();
        Manager manager = new Manager();
        Director director = new Director();
        CFO cfo = new CFO();

        teamLead.setNext(manager).setNext(director).setNext(cfo);

        List<ExpenseRequest> requests = List.of(
            new ExpenseRequest(500, "Office supplies"),
            new ExpenseRequest(5_000, "Team offsite"),
            new ExpenseRequest(30_000, "New servers"),
            new ExpenseRequest(100_000, "Acquisition deal")
        );

        for (ExpenseRequest req : requests) {
            System.out.println(teamLead.handle(req));
        }
    }
}
